subprojects/core/src/main/java/org/gradle/internal/service/scopes/ExecutionBuildServices.java

```diff
@@ -208,10 +215,11 @@ public ExecutionEngine createExecutionEngine(
             new IdentifyStep<>(buildOperationRunner,
             new IdentityCacheStep<>(buildOperationProgressEventEmitter,
             new ExecuteWorkBuildOperationFiringStep<>(buildOperationRunner,
+            new CheckDependenciesStep<>(problemHandler,
```

Comment:
💭 This is the "entry point" of the change: we now have the `CheckDependenciesStep` run early, regardless if we might later skip execution for `NO-SOURCE`. This fixes the problems.